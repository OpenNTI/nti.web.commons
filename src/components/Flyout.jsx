import React, {PropTypes} from 'react';
import ReactDOM from 'react-dom';
import cx from 'classnames';

import {getEffectiveZIndex, getViewportHeight, getViewportWidth, getScrollParent} from 'nti-lib-dom';

const {createElement: ce} = global.document || {};
const makeDOM = o => ce && Object.assign(ce.call(document, o.tag || 'div'), o);

function getRect (el) {
	const p = e => e && e.offsetParent;
	const r = e => p(e) ? [e].concat(r(p(e))) : [e];
	const sum = (a, e) => (a.top += e.offsetTop, a.left += e.offsetLeft, a);

	return r(el).reduce(sum, {
		left: 0,
		top: 0,
		width: el.offsetWidth,
		height: el.offsetHeight,
		get right () { return this.left + this.width; },
		get bottom () { return this.top + this.height; }
	});
}

function getAlignments (str) {
	str = str.toLowerCase();
	const [vert, horz] = str.split('-');

	const VALID_VERT = /^top|middle|bottom$/;
	const VALID_HORZ = /^left|center|right$/;

	return {
		horizontal: VALID_HORZ.test(horz) ? horz : 'left',
		vertical: VALID_VERT.test(vert) ? vert : 'bottom'
	};
}


function isFixed (el) {
	const hasFixedPosition = x => x && getComputedStyle(x).position === 'fixed';

	while(el && !hasFixedPosition(el)) {
		el = el.offsetParent;
	}

	return hasFixedPosition(el);
}


const ALIGNERS = {

	top (dim, {top}) {
		return {
			side: 'top',
			top: Math.round(top - dim) + 'px'
		};
	},

	middle (dim, {top, height}) {
		return {
			side: 'middle',
			top: Math.round((top + (height / 2)) - (dim / 2)) + 'px'
		};
	},

	bottom (dim, {bottom}) {
		return {
			side: 'bottom',
			top: Math.round(bottom) + 'px'
		};
	},



	left (dim, {left}) {
		return {
			side: 'left',
			left: Math.round(left) + 'px'
		};
	},

	center (dim, {left, width}) {
		return {
			left: Math.round((left + (width / 2)) - (dim / 2)) + 'px',
			side: 'center'
		};
	},

	right (dim, {left, width}) {
		return {
			side: 'right',
			left: Math.round((left + width) - dim) + 'px'
		};
	}
};


export default class Flyout extends React.Component {

	static propTypes = {
		trigger: PropTypes.any,
		children: PropTypes.any,
		className: PropTypes.string,
		alignment: PropTypes.string,
		afterAlign: PropTypes.func,
		onDismiss: PropTypes.func,
		arrow: PropTypes.bool
	}

	static defaultProps = {
		alignment: 'bottom center'
	}

	constructor (props) {
		super(props);
		this.state = {alignment: {}};
		this.fly = makeDOM({className: cx('fly-wrapper', props.className)});
		this.onToggle = this.onToggle.bind(this);
		this.attachFlyoutRef = this.attachFlyoutRef.bind(this);
		this.maybeDismiss = this.maybeDismiss.bind(this);

		this.realign = () => {
			clearTimeout(this.realign.timeout);
			this.realign.timeout = setTimeout(()=> this.align(), 50);
		};
	}


	get trigger () {
		return ReactDOM.findDOMNode(this);
	}


	componentDidMount () {
		document.body.appendChild(this.fly);
	}


	componentWillUnmount () {
		ReactDOM.unmountComponentAtNode(this.fly);
		document.body.removeChild(this.fly);
	}


	componentWillReceiveProps (nextProps) {
		if (nextProps.className !== this.props.className) {
			this.fly.className = cx('fly-wrapper', nextProps.className);
		}
	}


	componentDidUpdate (prevProps, prevState) {
		const {props: {onDismiss}, state: {open}} = this;
		const {open: wasOpen} = prevState;

		if (open) {
			this.renderFlyout();
		} else {
			ReactDOM.unmountComponentAtNode(this.fly);
		}

		if (wasOpen && !open && onDismiss) {
			onDismiss();
		}
	}


	maybeDismiss (e, cb) {
		const {target} = e || {};
		const {trigger, flyout, state: {open}} = this;
		const finish = () => typeof cb === 'function' && cb();
		const inDom = document.contains(target); // If the target was removed from dom after click or rerender it will not be in the dom anymore.

		if (e && (!trigger || !flyout || !open || !inDom || target === trigger || trigger.contains(target))) {
			return finish();
		}

		if (!e || (!flyout.contains(target) && flyout !== target)) {
			this.dismiss(finish);
		} else {
			finish();
		}
	}


	dismiss (cb) {
		this.setState({
			open: false,
			aligning: true
		}, () => {
			if (typeof cb === 'function') {
				cb();
			}
		});
	}


	listenToScroll (scroller) {
		const {scrollparent: old} = this;
		if (old === scroller) {
			return;
		}

		if (old) {
			old.removeEventListener('scroll', this.realign);
		}

		delete this.scrollparent;

		if (scroller) {
			scroller.addEventListener('scroll', this.realign);
			this.scrollparent = scroller;
		}
	}


	attachFlyoutRef (ref) {

		if (ref && !this.flyout) {
			window.addEventListener('resize', this.realign);
			window.document.addEventListener('click', this.maybeDismiss);
			this.listenToScroll(getScrollParent(ref) || window);
		} else if (!ref) {
			window.removeEventListener('resize', this.realign);
			window.document.removeEventListener('click', this.maybeDismiss);
			this.listenToScroll(null);
		}

		this.flyout = ref;

		if (!ref) {return;}

		this.align();
	}


	align (cb = this.props.afterAlign, noRetry = false) {
		const rect = getRect(this.trigger);
		const {offsetWidth: width, offsetHeight: height} = this.flyout || {};

		function flip (a) {
			return ({
				center: 'left',
				middle: 'bottom',
				left: 'right',
				top: 'bottom'
			})[a] || a;
		}

		const finish = () => {
			this.setState(
				{aligning: false},
				typeof cb === 'function' ? cb : void cb);
		};

		const calculateAlignment = (alignments, attempts = 1) => {
			const y = ALIGNERS[alignments.vertical](height, rect);
			const x = ALIGNERS[alignments.horizontal](width, rect);

			const alignment = {...y, ...x, dimensions: {height, width}};

			alignment.side = {[x.side]: 1, [y.side]: 1};

			this.setState({alignment}, () => {
				const {top, left, right, bottom} = this.flyout.getBoundingClientRect();
				const {vertical, horizontal} = alignments;
				let realignments = alignments;

				if (left < 0 || right > getViewportWidth()) {
					realignments = Object.assign({}, alignments, {horizontal: flip(horizontal)});
				}

				if (top < 0 || bottom > getViewportHeight()) {
					realignments = Object.assign({}, alignments, {vertical: flip(vertical)});
				}

				if (noRetry || alignments === realignments || attempts >= 3) {
					finish();
				} else {
					calculateAlignment(realignments, attempts + 1);
				}
			});
		};

		if (!this.flyout) {
			return finish();
		}

		calculateAlignment(getAlignments(this.props.alignment));
	}


	onToggle (e, cb) {
		if (e) {
			if (e.isPropagationStopped()) {
				return;
			}

			e.preventDefault();
			e.stopPropagation();
		}

		const {open} = this.state;

		this.setState(
			{ open: !open, aligning: true },
			typeof cb === 'function' ? cb : void cb);
	}


	render () {
		const props = Object.assign({}, this.props, {children: void 0});
		let {trigger: Trigger} = this.props;

		if (!Trigger) {
			Trigger = ( <button>Trigger</button> );
		}

		if (React.isValidElement(Trigger)) {
			const {onClick} = Trigger.props;
			let onToggle = this.onToggle;
			if (onClick) {
				onToggle = e => {
					onClick(e);
					this.onToggle(e);
				};
			}

			return React.cloneElement(Trigger, {onClick: onToggle});
		}

		return (
			<Trigger {...props} onClick={this.onToggle}/>
		);
	}


	renderFlyout () {
		const {props: {children, className, arrow}, state: {aligning, alignment}} = this;
		const {trigger} = this;
		const fixed = isFixed(trigger);

		const style = {
			position: fixed ? 'fixed' : 'absolute',
			visibility: aligning ? 'hidden' : void 0,
			top: alignment.top,
			left: alignment.left,
			zIndex: getEffectiveZIndex(trigger) + 1
		};

		const css = cx('flyout', className, alignment.side, {fixed, arrow});

		const flyout = ReactDOM.render(
			<div className={css} ref={this.attachFlyoutRef} style={style}>
				{arrow && <div className="flyout-arrow"/>}
				{children}
			</div>
		, this.fly, () => {

			if (flyout && this.flyout === flyout) {
				const {offsetWidth: width, offsetHeight: height} = flyout;
				const {dimensions: dim} = alignment;
				if (dim && (dim.width !== width || dim.height !== height)) {
					this.realign();
				}
			}
		});
	}
}
