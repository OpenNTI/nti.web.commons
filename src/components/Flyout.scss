@import "~nti-style-common";

.flyout-wrapper {
	@include clearfix;
	margin: 0;
	padding: 0;
}

.flyout {

	&.arrow.bottom {
		margin-top: 1rem;
	}

	&.arrow.top {
		margin-top: -1rem;
	}



	&.right {
		.flyout-arrow {
			right: rem-calc(12px);
		}
	}

	&.left {
		.flyout-arrow {
			left: rem-calc(12px);
		}
	}

	&.top {
		.flyout-arrow {
			bottom: 0;
			transform: rotate(180deg);
		}
	}

	&.bottom {
		.flyout-arrow {
			top: 0;
		}
	}

	&.center {
		.flyout-arrow {
			left: 50%;
			margin-left: -7px;
		}
	}

	.flyout-arrow {
		position: absolute;
		clip: rect(-15px, 25px, 0px, -10px);
		display: block;
		padding: 0;
		margin: 0;

		&::after {
			content: '';
			box-shadow: 0 0 4px 0 rgba(0,0,0,0.21);
			width: 15px;
			height: 15px;
			background: white;
			transform: translateY(-50%) rotate(45deg);
			display: block;
		}
	}

	.flyout-fullwidth-btn {
		width: 100%;
		height: 2rem;
		border-radius: 0.25rem;
		color: white;
		text-align: center;
		vertical-align: middle;
		line-height: 2rem;
		font-weight: 100;
		font-size: 0.875rem;
		font-family: $body-font-family;
		cursor: pointer;
		background-color: #d1d1d1;

		&.changed {
			background-color: $primary-blue;
		}
	}
}
