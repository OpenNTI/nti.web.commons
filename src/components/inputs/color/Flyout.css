.color-flyout-trigger {
	display: inline-block;
	width: 25px;
	height: 25px;
	border-radius: 25px;
	border: 3px solid white;
	box-shadow: 0 0 3px 1px rgba(0, 0, 0, 0.3);
	cursor: pointer;
}

.color-flyout-picker {
	padding: 0.5rem;
}