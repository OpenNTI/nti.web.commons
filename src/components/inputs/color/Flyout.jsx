import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames/bind';

import * as Flyout from '../../../flyout';

import Styles from './Flyout.css';
import Hue from './Hue';
import SaturationBrightness from './saturation-brightness';
import Text from './text';
import Preset from './preset-swatches';

const cx = classnames.bind(Styles);

ColorFlyout.ALIGNMENTS = Flyout.ALIGNMENTS;
ColorFlyout.propTypes = {
	className: PropTypes.string,
	swatches: PropTypes.array,
	value: PropTypes.shape({
		hex: PropTypes.shape({
			toString: PropTypes.func
		})
	}),
	onChange: PropTypes.func
};
export default function ColorFlyout ({className, onChange, value, swatches, ...otherProps}) {
	const style = {
		background: value ? value.hex.toString() : '#fff'
	};

	const trigger = (
		<span className={cx('color-flyout-trigger', className)} role="button" style={style} />
	);

	return (
		<Flyout.Triggered
			{...otherProps}
			trigger={trigger}
		>
			<div className={cx('color-flyout-picker')}>
				<SaturationBrightness value={value} onChange={onChange} />
				<Hue value={value} onChange={onChange} />
				<Text value={value} onChange={onChange} />
				<Preset swatches={swatches} selected={value} onSelect={onChange} />
			</div>
		</Flyout.Triggered>
	);
}