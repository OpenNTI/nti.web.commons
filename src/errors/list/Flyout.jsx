import PropTypes from 'prop-types';
import classnames from 'classnames/bind';

import { scoped } from '@nti/lib-locale';

import { isWarning } from '../messages';
import { Link } from '../linking/';
import Text from '../../text';
import * as Icons from '../../icons';
import { Triggered } from '../../flyout';

import Styles from './Styles.css';

const cx = classnames.bind(Styles);

const t = scoped('web-commons.errors.list.Flyout', {
	errors: {
		one: '%(count)s Error',
		other: '%(count)s Errors',
	},
	warnings: {
		one: '%(count)s Warning',
		other: '%(count)s Warnings',
	},
});

ErrorListFlyout.propTypes = {
	errors: PropTypes.array,
};
export default function ErrorListFlyout({ errors }) {
	const types = errors.reduce(
		(acc, error) => {
			if (isWarning(error)) {
				acc.warnings.push(error);
			} else {
				acc.errors.push(error);
			}

			return acc;
		},
		{ errors: [], warnings: [] }
	);

	const labels = [
		<div className={cx('label')} key="errors">
			<Icons.Alert className={cx('icon')} />
			<Text.Base
				className={cx('errors')}
				getString={t}
				localeKey="errors"
				with={{ count: types.errors.length }}
			/>
		</div>,
	];

	if (types.warnings.length > 0) {
		labels.push(
			<div className={cx('label', 'warning')} key="warnings">
				<Icons.Alert className={cx('icon')} />
				<Text.Base
					className={cx('warnings')}
					getString={t}
					localeKey="warnings"
					with={{ count: types.warnings.length }}
				/>
			</div>
		);
	}

	const trigger = <div className={cx('error-flyout-trigger')}>{labels}</div>;
	const sorted = [...types.warnings, ...types.errors];

	return (
		<Triggered trigger={trigger} arrow>
			<ul className={cx('error-list')}>
				{sorted.map((err, index) => (
					<li key={`${index}`}>
						<Link className={cx('error-list-item')} error={err} />
					</li>
				))}
			</ul>
		</Triggered>
	);
}
