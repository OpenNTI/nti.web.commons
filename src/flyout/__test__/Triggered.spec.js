/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';

import Flyout from '../Triggered';

const spyOn = (...args) => jest.spyOn(...args);

//We cannot currently get around not using this method...
//there are other test utils that allow better component access to tests.
const getDom = cmp => ReactDOM.findDOMNode(cmp); //eslint-disable-line
const getText = cmp => getDom(cmp).textContent;

const render = (node, cmp, props = {}, ...children) => new Promise(next => {
	let ref;
	ReactDOM.render(
		React.createElement(cmp, {...props, ref (x) {ref = x; props.ref && props.ref(x);}}, ...children),
		node,
		() => next(ref)
	);
});

describe('Triggered Flyout', () => {
	let container = document.createElement('div');
	let newNode;

	document.body.appendChild(container);

	function setup () {
		beforeEach(()=> {
			if (newNode) {
				document.body.removeChild(newNode);
			}
			newNode = document.createElement('div');
			document.body.appendChild(newNode);
		});

		afterEach(()=> {
			if (newNode) {
				ReactDOM.unmountComponentAtNode(newNode);
				document.body.removeChild(newNode);
				newNode = null;
			}
		});
	}

	beforeAll(()=> {
		document.scrollingElement = window; //jest's jsdom env doesn't define this
		jest.restoreAllMocks();
	});

	afterAll(()=> {
		delete document.scrollingElement;
		ReactDOM.unmountComponentAtNode(container);
		document.body.removeChild(container);
	});

	const testRender = (props, ...children) => [
		render(newNode, Flyout, props, ...children),
		render(container, Flyout, props, ...children)
	];

	setup();

	test('Base Case', async () => {
		const cmps = await Promise.all(testRender());

		cmps.map(getText)
			.forEach(x => expect(x).toEqual('Trigger'));
	});

	test('Base Case: Specify Trigger', async () => {
		const value = 'Test';

		const cmps = await Promise.all(testRender({trigger: 'input', type: 'button', value}));

		cmps.map(getDom)
			.forEach(x => {
				expect(x.tagName).toEqual('INPUT');
				expect(x.value).toEqual(value);
			});
	});

	test('Base Case: Specify Element Trigger', async () => {
		const value = 'Test';
		const element = (
			<a href="#test" className="foobar">{value}</a>
		);

		const cmps = await Promise.all(testRender({trigger: element}));

		cmps.map(getDom)
			.forEach(x => {
				expect(x.tagName).toEqual('A');
				expect(x.getAttribute('href')).toEqual('#test');
				expect(x.classList.contains('foobar')).toBeTruthy();
				expect(x.textContent).toEqual(value);
			});
	});

	test('Opening the flyout should add listeners to window, and document', async () => {
		spyOn(window, 'addEventListener');
		spyOn(window, 'removeEventListener');
		spyOn(window.document, 'addEventListener');
		spyOn(window.document, 'removeEventListener');

		let step = null;

		function afterAlign () {
			step();
		}

		const [component] = await Promise.all(testRender({afterAlign}, <div>Foobar</div>));

		await new Promise(next =>{
			step = next;
			component.onToggle();
		});

		expect(window.addEventListener).toHaveBeenCalledWith('resize', component.realign);
		// expect(window.addEventListener).toHaveBeenCalledWith('scroll', component.realign);
		expect(window.document.addEventListener).toHaveBeenCalledWith('click', component.maybeDismiss);
	});

	test('Closing the flyout should remove listeners to window, and document', async () => {
		spyOn(window, 'addEventListener');
		spyOn(window, 'removeEventListener');
		spyOn(window.document, 'addEventListener');
		spyOn(window.document, 'removeEventListener');

		let step = null;

		function afterAlign () {
			step();
		}

		const [component] = await Promise.all(testRender({afterAlign}, <div>Foobar2</div>));

		await new Promise(next =>{
			step = next;
			component.onToggle();
		});

		await new Promise(next =>{
			component.onToggle(null, next);
		});

		expect(window.removeEventListener).toHaveBeenCalledWith('resize', expect.any(Function));
		// expect(window.removeEventListener).toHaveBeenCalledWith('scroll', expect.any(Function));
		expect(window.document.removeEventListener).toHaveBeenCalledWith('click', expect.any(Function));
	});

	test('Flys echo classnames', async () => {
		const cmps = await Promise.all(testRender({className: 'awesome sauce'}, <div>Lala</div> ));
		await Promise.all(cmps.map(x => new Promise(next => x.onToggle(null, next))));

		for (let c of cmps) {
			expect(c.state.open).toBeTruthy();
			expect(c.fly.className).toEqual('fly-wrapper awesome sauce');
			expect(c.flyout.className).toEqual('flyout awesome sauce bottom center');
			expect(c.flyout.textContent).toEqual('Lala');
			expect(c.fly.contains(c.flyout)).toBeTruthy();
		}

	});
});
