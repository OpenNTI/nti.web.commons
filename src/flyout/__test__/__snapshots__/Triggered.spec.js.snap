// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Triggered Flyout [Non-forwarded-refs] Component trigger.getDOMNode() returns node 1`] = `
<a
  href="#test"
>
  Test
</a>
`;

exports[`Triggered Flyout [Non-forwarded-refs] Warns for Component triggers that do not implement getDOMNode() 1`] = `
<a
  href="#test"
>
  Test
</a>
`;

exports[`Triggered Flyout [Non-forwarded-refs] Warns for Component triggers that do not implement getDOMNode() 2`] = `
<a
  href="#test"
>
  Test
</a>
`;

exports[`Triggered Flyout [Non-forwarded-refs] Warns for Stateless component triggers 1`] = `
<a
  href="#test"
>
  Test
</a>
`;

exports[`Triggered Flyout [Non-forwarded-refs] Warns for Stateless component triggers 2`] = `
<a
  href="#test"
>
  Test
</a>
`;

exports[`Triggered Flyout Base Case 1`] = `
<button
  className="flyout-trigger flyout-closed"
  onClick={[Function]}
>
  Trigger
</button>
`;

exports[`Triggered Flyout Children recieve onDismiss prop 1`] = `
Array [
  <button
    className="flyout-trigger flyout-open"
  >
    Trigger
  </button>,
  <div
    data-target-tag-name="DIV"
    data-type="portal"
  >
    <div
      className="flyout bottom center opened"
      style={
        Object {
          "left": "0px",
          "position": "absolute",
          "top": "0px",
          "visibility": undefined,
          "zIndex": undefined,
        }
      }
    >
      <div
        className="flyout-inner"
        style={
          Object {
            "maxHeight": null,
            "maxWidth": null,
          }
        }
      >
        <div />
      </div>
    </div>
  </div>,
]
`;

exports[`Triggered Flyout Closing the flyout should remove listeners to window, and document 1`] = `
<button
  className="flyout-trigger flyout-closed"
  onClick={[Function]}
>
  Trigger
</button>
`;

exports[`Triggered Flyout Controlled 1`] = `
<button
  className="flyout-trigger flyout-closed"
>
  Trigger
</button>
`;

exports[`Triggered Flyout Controlled 2`] = `
Array [
  <button
    className="flyout-trigger flyout-open"
  >
    Trigger
  </button>,
  <div
    data-target-tag-name="DIV"
    data-type="portal"
  >
    <div
      className="flyout bottom center flyout-arrow opened"
      style={
        Object {
          "left": "0px",
          "position": "absolute",
          "top": "15px",
          "visibility": undefined,
          "zIndex": undefined,
        }
      }
    >
      <div
        className="flyout-arrow"
      />
      <div
        className="flyout-inner"
        style={
          Object {
            "maxHeight": null,
            "maxWidth": null,
          }
        }
      >
        <div>
          My Flyout Content
        </div>
        <div>
          My New Flyout Content
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Triggered Flyout Controlled 3`] = `
<div
  className="flyout-trigger new flyout-closed"
/>
`;

exports[`Triggered Flyout Controlled doesn't auto dismiss 1`] = `
Array [
  <div
    className="flyout-trigger flyout-open"
  />,
  <div
    data-target-tag-name="DIV"
    data-type="portal"
  >
    <div
      className="flyout bottom center opened"
      style={
        Object {
          "left": "0px",
          "position": "absolute",
          "top": "0px",
          "visibility": undefined,
          "zIndex": undefined,
        }
      }
    >
      <div
        className="flyout-inner"
        style={
          Object {
            "maxHeight": null,
            "maxWidth": null,
          }
        }
      >
        <div>
          Flyout Content
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Triggered Flyout Flys echo classnames 1`] = `
<button
  className="flyout-trigger awesome sauce flyout-closed"
  onClick={[Function]}
>
  Trigger
</button>
`;

exports[`Triggered Flyout Flys echo classnames 2`] = `
Array [
  <button
    className="flyout-trigger awesome sauce flyout-open"
    onClick={[Function]}
  >
    Trigger
  </button>,
  <div
    data-target-tag-name="DIV"
    data-type="portal"
  >
    <div
      className="flyout awesome sauce bottom center opened"
      style={
        Object {
          "left": "0px",
          "position": "absolute",
          "top": "0px",
          "visibility": undefined,
          "zIndex": undefined,
        }
      }
    >
      <div
        className="flyout-inner"
        style={
          Object {
            "maxHeight": null,
            "maxWidth": null,
          }
        }
      >
        <div>
          Lala
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Triggered Flyout Opening the flyout should add listeners to window, and document 1`] = `
<button
  className="flyout-trigger flyout-closed"
  onClick={[Function]}
>
  Trigger
</button>
`;

exports[`Triggered Flyout Realigns when flyout size changes 1`] = `
Array [
  <button
    className="flyout-trigger flyout-open"
  >
    Trigger
  </button>,
  <div
    data-target-tag-name="DIV"
    data-type="portal"
  >
    <div
      className="flyout bottom center opened"
      style={
        Object {
          "left": "-50px",
          "position": "absolute",
          "top": "0px",
          "visibility": undefined,
          "zIndex": undefined,
        }
      }
    >
      <div
        className="flyout-inner"
        style={
          Object {
            "maxHeight": null,
            "maxWidth": null,
          }
        }
      >
        <div
          style={
            Object {
              "height": 100,
              "width": 100,
            }
          }
        />
      </div>
    </div>
  </div>,
]
`;

exports[`Triggered Flyout Ref is not dropped from custom trigger 1`] = `
<a
  className="flyout-trigger flyout-closed"
  href="#test"
  onClick={[Function]}
  value="Test"
>
  Test
</a>
`;

exports[`Triggered Flyout Specify Element Trigger 1`] = `
<a
  className="flyout-trigger flyout-closed"
  href="#test"
  onClick={[Function]}
  value="Test"
>
  Test
</a>
`;

exports[`Triggered Flyout Specify Trigger 1`] = `
<button
  className="flyout-trigger flyout-closed"
  onClick={[Function]}
  type="button"
  value="Test"
>
  input
</button>
`;

exports[`Triggered Flyout Transition Support 1`] = `
<button
  className="flyout-trigger flyout-closed"
  onClick={[Function]}
>
  Trigger
</button>
`;

exports[`Triggered Flyout Transition Support 2`] = `
Array [
  <button
    className="flyout-trigger flyout-open"
    onClick={[Function]}
  >
    Trigger
  </button>,
  <div
    data-target-tag-name="DIV"
    data-type="portal"
  >
    <div
      className="flyout fade bottom center opening"
      style={
        Object {
          "left": "0px",
          "position": "absolute",
          "top": "0px",
          "visibility": undefined,
          "zIndex": undefined,
        }
      }
    >
      <div
        className="flyout-inner"
        style={
          Object {
            "maxHeight": null,
            "maxWidth": null,
          }
        }
      />
    </div>
  </div>,
]
`;

exports[`Triggered Flyout Transition Support 3`] = `
Array [
  <button
    className="flyout-trigger flyout-open"
    onClick={[Function]}
  >
    Trigger
  </button>,
  <div
    data-target-tag-name="DIV"
    data-type="portal"
  >
    <div
      className="flyout fade bottom center opened"
      style={
        Object {
          "left": "0px",
          "position": "absolute",
          "top": "0px",
          "visibility": undefined,
          "zIndex": undefined,
        }
      }
    >
      <div
        className="flyout-inner"
        style={
          Object {
            "maxHeight": null,
            "maxWidth": null,
          }
        }
      />
    </div>
  </div>,
]
`;

exports[`Triggered Flyout Transition Support 4`] = `
Array [
  <button
    className="flyout-trigger flyout-open"
    onClick={[Function]}
  >
    Trigger
  </button>,
  <div
    data-target-tag-name="DIV"
    data-type="portal"
  >
    <div
      className="flyout fade bottom center closing"
      style={
        Object {
          "left": "0px",
          "position": "absolute",
          "top": "0px",
          "visibility": undefined,
          "zIndex": undefined,
        }
      }
    >
      <div
        className="flyout-inner"
        style={
          Object {
            "maxHeight": null,
            "maxWidth": null,
          }
        }
      />
    </div>
  </div>,
]
`;

exports[`Triggered Flyout Transition Support 5`] = `
<button
  className="flyout-trigger flyout-closed"
  onClick={[Function]}
>
  Trigger
</button>
`;

exports[`Triggered Flyout Trigger Click Handlers 1`] = `
<div
  className="flyout-trigger custom-trigger flyout-closed"
  onClick={[Function]}
/>
`;

exports[`Triggered Flyout alignment shorts if closed 1`] = `
<button
  className="flyout-trigger flyout-closed"
  onClick={[Function]}
>
  Trigger
</button>
`;

exports[`Triggered Flyout constrained alignment does not limit height for non-top-aligned, non-fixed flyouts 1`] = `
Array [
  <button
    className="flyout-trigger flyout-open"
  >
    Trigger
  </button>,
  <div
    data-target-tag-name="DIV"
    data-type="portal"
  >
    <div
      className="flyout bottom center opened"
      style={
        Object {
          "left": "0px",
          "position": "absolute",
          "top": "0px",
          "visibility": undefined,
          "zIndex": undefined,
        }
      }
    >
      <div
        className="flyout-inner"
        style={
          Object {
            "maxHeight": null,
            "maxWidth": "1024px",
          }
        }
      >
        <div />
      </div>
    </div>
  </div>,
]
`;

exports[`Triggered Flyout defaultState 1`] = `
Array [
  <button
    className="flyout-trigger flyout-open"
    onClick={[Function]}
  >
    Trigger
  </button>,
  <div
    data-target-tag-name="DIV"
    data-type="portal"
  >
    <div
      className="flyout bottom center opened"
      style={
        Object {
          "left": "0px",
          "position": "absolute",
          "top": "0px",
          "visibility": undefined,
          "zIndex": undefined,
        }
      }
    >
      <div
        className="flyout-inner"
        style={
          Object {
            "maxHeight": null,
            "maxWidth": null,
          }
        }
      >
        <div>
          My Flyout Content
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Triggered Flyout getViewportRelativeAlignments 1`] = `
Object {
  "bottom": 68,
  "height": 600,
  "left": 50,
  "right": 774,
  "top": 100,
  "width": 200,
}
`;

exports[`Triggered Flyout getViewportRelativeAlignments 2`] = `
Object {
  "height": 600,
  "width": 200,
}
`;

exports[`Triggered Flyout getViewportRelativeAlignments 3`] = `
Object {
  "height": 600,
  "left": 50,
  "top": 100,
  "width": 200,
}
`;

exports[`Triggered Flyout getViewportRelativeAlignments 4`] = `
Object {
  "bottom": 68,
  "height": 600,
  "right": 774,
  "width": 200,
}
`;

exports[`Triggered Flyout getViewportRelativeAlignments 5`] = `
Object {
  "height": 600,
  "left": 50,
  "right": 774,
  "top": 100,
  "width": 200,
}
`;

exports[`Triggered Flyout getViewportRelativeAlignments 6`] = `
Object {
  "bottom": 68,
  "height": 600,
  "left": 50,
  "right": 774,
  "width": 200,
}
`;

exports[`Triggered Flyout getViewportRelativeAlignments 7`] = `
Object {
  "bottom": 68,
  "height": 600,
  "left": 50,
  "top": 100,
  "width": 200,
}
`;

exports[`Triggered Flyout getViewportRelativeAlignments 8`] = `
Object {
  "bottom": 68,
  "height": 600,
  "right": 774,
  "top": 100,
  "width": 200,
}
`;

exports[`Triggered Flyout hover over trigger opens flyout 1`] = `
<button
  className="flyout-trigger flyout-closed"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Trigger
</button>
`;

exports[`Triggered Flyout hover over trigger opens flyout 2`] = `
Object {
  "closeTimeout": 500,
  "openTimeout": 300,
}
`;

exports[`Triggered Flyout hover over trigger opens flyout 3`] = `
Object {
  "closeTimeout": 2,
  "openTimeout": 1,
}
`;

exports[`Triggered Flyout maybeDismiss() 1`] = `
Array [
  <button
    className="flyout-trigger flyout-open"
    onClick={[Function]}
  >
    Trigger
  </button>,
  <div
    data-target-tag-name="DIV"
    data-type="portal"
  >
    <div
      className="flyout bottom center opened"
      style={
        Object {
          "left": "0px",
          "position": "absolute",
          "top": "0px",
          "visibility": undefined,
          "zIndex": undefined,
        }
      }
    >
      <div
        className="flyout-inner"
        style={
          Object {
            "maxHeight": null,
            "maxWidth": null,
          }
        }
      >
        <div />
      </div>
    </div>
  </div>,
]
`;

exports[`Triggered Flyout mouse out of flyout closes 1`] = `
<button
  className="flyout-trigger flyout-closed"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Trigger
</button>
`;

exports[`Triggered Flyout position fixed resolition 1`] = `
Array [
  <button
    className="flyout-trigger flyout-open"
  >
    Trigger
  </button>,
  <div
    data-target-tag-name="DIV"
    data-type="portal"
  >
    <div
      className="flyout bottom center is-fixed opened"
      style={
        Object {
          "left": "0px",
          "position": "fixed",
          "top": "0px",
          "visibility": undefined,
          "zIndex": 11,
        }
      }
    >
      <div
        className="flyout-inner"
        style={
          Object {
            "maxHeight": "768px",
            "maxWidth": "1024px",
          }
        }
      >
        <div>
          My Flyout Content
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Triggered Flyout scrolling realigns 1`] = `
Array [
  <button
    className="flyout-trigger flyout-open"
  >
    Trigger
  </button>,
  <div
    data-target-tag-name="DIV"
    data-type="portal"
  >
    <div
      className="flyout top center opened"
      style={
        Object {
          "bottom": "NaNpx",
          "left": "NaNpx",
          "position": "absolute",
          "visibility": undefined,
          "zIndex": undefined,
        }
      }
    >
      <div
        className="flyout-inner"
        style={
          Object {
            "maxHeight": null,
            "maxWidth": null,
          }
        }
      />
    </div>
  </div>,
]
`;

exports[`Triggered Flyout zIndex resolition 1`] = `
Array [
  <button
    className="flyout-trigger flyout-open"
  >
    Trigger
  </button>,
  <div
    data-target-tag-name="DIV"
    data-type="portal"
  >
    <div
      className="flyout top center opened"
      style={
        Object {
          "bottom": "NaNpx",
          "left": "NaNpx",
          "position": "absolute",
          "visibility": undefined,
          "zIndex": 11,
        }
      }
    >
      <div
        className="flyout-inner"
        style={
          Object {
            "maxHeight": null,
            "maxWidth": "1024px",
          }
        }
      >
        <div>
          My Flyout Content
        </div>
      </div>
    </div>
  </div>,
]
`;
